### ISSUE TITLE
`map` and `apply` expressions don't work with server resources

### ISSUE THREAD

**Issue by kwmsmith** [blaze/blaze#1422]
iris.yml
```
iris:
    source: /path/to/blaze/examples/data/iris.csv
```

```python
In [1]: import blaze as bz

In [3]: bs = bz.Data('blaze://localhost:6363')

In [4]: bs
Out[4]:
Data:       <blaze.server.client.Client object at 0x11a4df750>
DataShape:  {
  iris: var * {
    sepal_length: float64,
    sepal_width: float64,
    petal_length: float64,
    petal_width: float64,
    species: ?string
  ...

In [6]: bz.compute(bs.iris.species.map(len, schema='int'))
---------------------------------------------------------------------------
NotImplementedError                       Traceback (most recent call last)
<ipython-input-6-2b12b14b6069> in <module>()
----> 1 bz.compute(bs.iris.species.map(len, schema='int'))

/Users/ksmith/anaconda/envs/blaze-3.4/lib/python3.4/site-packages/multipledispatch-0.4.8-py3.4.egg/multipledispatch/dispatcher.py in __call__(self, *args, **kwargs)
    162             self._cache[types] = func
    163         try:
--> 164             return func(*args, **kwargs)
    165
    166         except MDNotImplementedError:

/Users/ksmith/work/blaze/blaze-repo/blaze/interactive.py in compute(expr, **kwargs)
    166         raise ValueError("No data resources found")
    167     else:
--> 168         return compute(expr, resources, **kwargs)
    169
    170

/Users/ksmith/anaconda/envs/blaze-3.4/lib/python3.4/site-packages/multipledispatch-0.4.8-py3.4.egg/multipledispatch/dispatcher.py in __call__(self, *args, **kwargs)
    162             self._cache[types] = func
    163         try:
--> 164             return func(*args, **kwargs)
    165
    166         except MDNotImplementedError:

/Users/ksmith/work/blaze/blaze-repo/blaze/compute/core.py in compute(expr, d, **kwargs)
    404         d4 = d3
    405
--> 406     result = top_then_bottom_then_top_again_etc(expr3, d4, **kwargs)
    407     if post_compute_:
    408         result = post_compute_(expr3, result, scope=d4)

/Users/ksmith/work/blaze/blaze-repo/blaze/compute/core.py in top_then_bottom_then_top_again_etc(expr, scope, **kwargs)
    194         raise NotImplementedError("Don't know how to compute:\n"
    195                                   "expr: %s\n"
--> 196                                   "data: %s" % (expr3, scope4))
    197     else:
    198         return top_then_bottom_then_top_again_etc(expr3, scope4, **kwargs)

NotImplementedError: Don't know how to compute:
expr: Map(_child=_2.iris.species, func=len, _asschema='int', _name0=None)
data: {_2: <blaze.server.client.Client object at 0x11a4df750>}
```

**Comment by kwmsmith** [comment]
Not unexpected, the error comes when client.py tries to `json.dumps()` a function object.

One fix would be to generalize the json serialization to handle function objects.

Using `pickle` would require the function to be defined or available both on the client and server.

Using `dill` would essentially redefine the function on the server, so would not require the function to be defined there; dill isn't widely used and would introduce another external dependency, however.

For now, `map` and `apply` expressions can't work in a client-server setting.

**Comment by llllllllll** [comment]
Does this work with the `pickle` serialization format option? I am also okay with not supporting this because of the security implications of sending arbitrary functions. We might want to add a dispatch in `json_dumps` on object that raises a `ValueError` instead of a `NotImplementedErro` so that it can say `'Don't know how to serialize object of type %r: %r' % (type(ob).__name__, ob)`.

As a side note, I highly recommend not using the json serialization format for anything other than debugging or talking to some web based frontend because the performance is pretty bad with our custom object hooks.

**Comment by kwmsmith** [comment]
Closed by #1504
